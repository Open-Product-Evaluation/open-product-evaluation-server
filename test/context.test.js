const users = require('../seeds/data/user/user')
const surveys = require('../seeds/data/survey/survey')
const questions = require('../seeds/data/question/question')
const contexts = require('../seeds/data/context/context')
const { seedDatabase } = require('mongo-seeding')
const config = require('../config')
const request = require('./requesthelper')
const { getSeedID } = require('./helpers')

/* Functions for Querys */

function createContextQuery(name) {
  return {
    query: `mutation {
      createContext(data: {name: "${name}"}) {
        context {
          name
          activeQuestion{
            description
          }
          activeSurvey{
            title
          }
          owners {
            email
          }
          devices {
            name
          }
          states {
            key
          }
        }
      }
    }`,
  }
}

function updateContextQuery(contextID, contextName, activeQuestion, activeSurvey, owners) {
  return {
    query: `mutation {
      updateContext(contextID: "${contextID}", data: {name: "${contextName}", activeQuestion: "${activeQuestion}", activeSurvey: "${activeSurvey}", owners: ${JSON.stringify(owners)}}) {
        context {
          name
          activeQuestion{
            description
          }
          activeSurvey{
            title
          }
          owners {
            email
          }
          devices {
            name
          }
          states {
            key
          }
        }
      }
    }`,
  }
}

function deleteContextQuery(contextID) {
  return {
    query: `mutation {
      deleteContext(contextID: "${contextID}") {
        success
      }
    }`,
  }
}

function contextsQuery() {
  return {
    query: `{
      contexts {
        name
        activeQuestion{
          description
        }
        activeSurvey{
          title
        }
        owners {
          email
        }
        devices {
          name
        }
        states {
          key
        }
      }
    }
    `,
  }
}

function contextQuery(contextID) {
  return {
    query: `{
      context(contextID: "${contextID}") {
        name
        activeQuestion{
          description
        }
        activeSurvey{
          title
        }
        owners {
          email
        }
        devices {
          name
        }
        states {
          key
        }
      }
    }`,
  }
}

/* Tests */

describe('Context', () => {
  describe('Admin', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[1]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all contexts [Query]', async () => {
      const query = contextsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return context owned by User [Query]', async () => {
      const context = contexts[0]
      const query = contextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return context not owned by User [Query]', async () => {
      const context = contexts[1]
      const query = contextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should create context [Mutation]', async () => {
      const query = createContextQuery('TestContext')
      const res = await request.user(query, jwtToken)
      const { data, errors } = res
      expect(data).toMatchSnapshot()
      expect(errors).toBeUndefined()
    })
    it('should update context owned by User [Mutation]', async () => {
      const context = contexts[1]
      const question = questions[0]
      const survey = surveys[0]
      const user = users[0]
      const query = updateContextQuery(getSeedID(context), 'RenamedTestContext', getSeedID(question), getSeedID(survey), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should update context not owned by User [Mutation]', async () => {
      const context = contexts[0]
      const question = questions[0]
      const survey = surveys[0]
      const user = users[0]
      const query = updateContextQuery(getSeedID(context), 'RenamedTestContext', getSeedID(question), getSeedID(survey), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should delete context owned by user [Mutation]', async () => {
      const context = contexts[0]
      const query = deleteContextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteContext.success).toBe(true)
    })
    it('should delete context not owned by user [Mutation]', async () => {
      const context = contexts[1]
      const query = deleteContextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteContext.success).toBe(true)
    })
  })
  describe('User', async () => {
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const expected = users[0]
      const query = {
        query: `mutation {
          login(data: {email: "${expected.email}", password: "password"}) {
            token
            user {
              id
              firstName
              lastName
              email
            }
          }
        }`,
      }
      const { data, errors } = await request.anon(query)
      data.login.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { login: { token } } = data
      jwtToken = token
    })
    it('should return all contexts [Query]', async () => {
      const query = contextsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return context owned by User [Query]', async () => {
      const context = contexts[0]
      const query = contextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should return context not owned by User [Query]', async () => {
      const context = contexts[1]
      const query = contextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors.length).toBe(1)
      expect(errors[0].path).toEqual(['context', 'owners'])
      expect(data).toMatchSnapshot()
    })
    it('should create context [Mutation]', async () => {
      const query = createContextQuery('TestContext')
      const res = await request.user(query, jwtToken)
      const { data, errors } = res
      expect(data).toMatchSnapshot()
      expect(errors).toBeUndefined()
    })
    it('should update context owned by User [Mutation]', async () => {
      const context = contexts[0]
      const question = questions[0]
      const survey = surveys[0]
      const user = users[0]
      const query = updateContextQuery(getSeedID(context), 'RenamedTestContext', getSeedID(question), getSeedID(survey), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data).toMatchSnapshot()
    })
    it('should not update context not owned by User [Mutation]', async () => {
      const context = contexts[1]
      const question = questions[0]
      const survey = surveys[0]
      const user = users[0]
      const query = updateContextQuery(getSeedID(context), 'RenamedTestContext', getSeedID(question), getSeedID(survey), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should delete context owned by user [Mutation]', async () => {
      const context = contexts[0]
      const query = deleteContextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteContext.success).toBe(true)
    })
    it('should not delete context not owned by user [Mutation]', async () => {
      const context = contexts[1]
      const query = deleteContextQuery(getSeedID(context))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
  describe.skip('Device', async () => {
    // TODO Not testable without device login function
    let jwtToken = ''
    beforeAll(async () => {
      await seedDatabase(config.seeder)
      const query = {
        query: `mutation{createDevice(data:{name:"TestDevice"}){
          token
       }}`,
      }
      const { data, errors } = await request.anon(query)
      data.createDevice.token.should.be.a('string')
      expect(errors).toBeUndefined()
      const { createDevice: { token } } = data
      jwtToken = token
    })
    it('should return contexts assigned to [Query]', async () => {
      const query = contextsQuery()
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteContext.success).toBe(true)
    })
    it('should return context assigned to [Query]', async () => {
      const device = contexts[0]
      const query = contextQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(errors).toBeUndefined()
      expect(data.deleteContext.success).toBe(true)
    })
    it('should not return context not assigned to [Query]', async () => {
      const device = contexts[0]
      const context = contexts[0]
      const user = users[0]
      const query = updateContextQuery(getSeedID(device), 'RenamedTestDevice', getSeedID(context), [getSeedID(user)])
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
    it('should not delete device [Mutation]', async () => {
      const device = contexts[3]
      const query = deleteContextQuery(getSeedID(device))
      const { data, errors } = await request.user(query, jwtToken)
      expect(data).toBeNull()
      expect(errors.length).toBeGreaterThan(0)
    })
  })
})
