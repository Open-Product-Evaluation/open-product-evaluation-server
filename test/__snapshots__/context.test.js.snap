// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context Admin should create context [Mutation] 1`] = `
Object {
  "createContext": Object {
    "context": Object {
      "activeQuestion": null,
      "activeSurvey": null,
      "devices": null,
      "name": "TestContext",
      "owners": Array [
        Object {
          "email": "jane@doe.com",
        },
      ],
      "states": null,
    },
  },
}
`;

exports[`Context Admin should return all contexts [Query] 1`] = `
Object {
  "contexts": Array [
    Object {
      "activeQuestion": Object {
        "description": "Bitte zeigen Sie mit einem \\"Like\\", ob Ihnen dieses Objekt gefällt.",
      },
      "activeSurvey": Object {
        "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
      },
      "devices": Array [
        Object {
          "name": "Forum Fernseher",
        },
        Object {
          "name": "Forum Tablet",
        },
      ],
      "name": "Forum Gummersbach",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
    Object {
      "activeQuestion": Object {
        "description": null,
      },
      "activeSurvey": Object {
        "title": "Öffentliche Umfrage im Rahmen des Modules \\"Computerethik\\"",
      },
      "devices": Array [
        Object {
          "name": "Mensa Kiosk",
        },
        Object {
          "name": "Mensa Tablet",
        },
      ],
      "name": "TH-Koeln: Gummersbach Mensa",
      "owners": Array [
        Object {
          "email": "jane@doe.com",
        },
      ],
      "states": null,
    },
  ],
}
`;

exports[`Context Admin should return context not owned by User [Query] 1`] = `
Object {
  "context": Object {
    "activeQuestion": Object {
      "description": null,
    },
    "activeSurvey": Object {
      "title": "Öffentliche Umfrage im Rahmen des Modules \\"Computerethik\\"",
    },
    "devices": Array [
      Object {
        "name": "Mensa Kiosk",
      },
      Object {
        "name": "Mensa Tablet",
      },
    ],
    "name": "TH-Koeln: Gummersbach Mensa",
    "owners": Array [
      Object {
        "email": "jane@doe.com",
      },
    ],
    "states": null,
  },
}
`;

exports[`Context Admin should return context owned by User [Query] 1`] = `
Object {
  "context": Object {
    "activeQuestion": Object {
      "description": "Bitte zeigen Sie mit einem \\"Like\\", ob Ihnen dieses Objekt gefällt.",
    },
    "activeSurvey": Object {
      "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
    },
    "devices": Array [
      Object {
        "name": "Forum Fernseher",
      },
      Object {
        "name": "Forum Tablet",
      },
    ],
    "name": "Forum Gummersbach",
    "owners": Array [
      Object {
        "email": "john@doe.com",
      },
    ],
    "states": null,
  },
}
`;

exports[`Context Admin should update context not owned by User [Mutation] 1`] = `
Object {
  "updateContext": Object {
    "context": Object {
      "activeQuestion": Object {
        "description": null,
      },
      "activeSurvey": Object {
        "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
      },
      "devices": Array [
        Object {
          "name": "Forum Fernseher",
        },
        Object {
          "name": "Forum Tablet",
        },
      ],
      "name": "RenamedTestContext",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
  },
}
`;

exports[`Context Admin should update context owned by User [Mutation] 1`] = `
Object {
  "updateContext": Object {
    "context": Object {
      "activeQuestion": Object {
        "description": null,
      },
      "activeSurvey": Object {
        "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
      },
      "devices": Array [
        Object {
          "name": "Mensa Kiosk",
        },
        Object {
          "name": "Mensa Tablet",
        },
      ],
      "name": "RenamedTestContext",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
  },
}
`;

exports[`Context User should create context [Mutation] 1`] = `
Object {
  "createContext": Object {
    "context": Object {
      "activeQuestion": null,
      "activeSurvey": null,
      "devices": null,
      "name": "TestContext",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
  },
}
`;

exports[`Context User should return all contexts [Query] 1`] = `
Object {
  "contexts": Array [
    Object {
      "activeQuestion": Object {
        "description": "Bitte zeigen Sie mit einem \\"Like\\", ob Ihnen dieses Objekt gefällt.",
      },
      "activeSurvey": Object {
        "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
      },
      "devices": Array [
        Object {
          "name": "Forum Fernseher",
        },
        Object {
          "name": "Forum Tablet",
        },
      ],
      "name": "Forum Gummersbach",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
  ],
}
`;

exports[`Context User should return context not owned by User [Query] 1`] = `
Object {
  "context": Object {
    "activeQuestion": Object {
      "description": null,
    },
    "activeSurvey": Object {
      "title": "Öffentliche Umfrage im Rahmen des Modules \\"Computerethik\\"",
    },
    "devices": Array [
      Object {
        "name": "Mensa Kiosk",
      },
      Object {
        "name": "Mensa Tablet",
      },
    ],
    "name": "TH-Koeln: Gummersbach Mensa",
    "owners": null,
    "states": null,
  },
}
`;

exports[`Context User should return context owned by User [Query] 1`] = `
Object {
  "context": Object {
    "activeQuestion": Object {
      "description": "Bitte zeigen Sie mit einem \\"Like\\", ob Ihnen dieses Objekt gefällt.",
    },
    "activeSurvey": Object {
      "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
    },
    "devices": Array [
      Object {
        "name": "Forum Fernseher",
      },
      Object {
        "name": "Forum Tablet",
      },
    ],
    "name": "Forum Gummersbach",
    "owners": Array [
      Object {
        "email": "john@doe.com",
      },
    ],
    "states": null,
  },
}
`;

exports[`Context User should update context owned by User [Mutation] 1`] = `
Object {
  "updateContext": Object {
    "context": Object {
      "activeQuestion": Object {
        "description": null,
      },
      "activeSurvey": Object {
        "title": "Untersuchung zum Verhalten von Informatikern in Hochschulen",
      },
      "devices": Array [
        Object {
          "name": "Forum Fernseher",
        },
        Object {
          "name": "Forum Tablet",
        },
      ],
      "name": "RenamedTestContext",
      "owners": Array [
        Object {
          "email": "john@doe.com",
        },
      ],
      "states": null,
    },
  },
}
`;
